### ISSUE TITLE
🐛 Files with `fn(() => (aborted = true));`  cause - entered unreachable code - when `--apply-unsafe`

### ISSUE THREAD

**Issue by mrazorvin** [rome/tools#4464]
### Environment information

```block
Rome: v12.0.0
```

### What happened?

command `./node_modules/.bin/rome check ./packages --apply-unsafe`

```js
  res.onAborted(() => (aborted = true));
```

cause entered unreachable code

```
Source Location: crates/rome_js_syntax/src/expr_ext.rs:489:18
Thread Name: rome::worker_5
Message: internal error: entered unreachable code
```

### Expected result

Fixed XXXX file(s) in YYYYms

### Code of Conduct

- [X] I agree to follow Rome's Code of Conduct

**Comment by ematipico** [comment]
I believe this issue was fixed. Could you try one of our latest nightlies? We're going to do a new release soon.

**Comment by denbezrukov** [comment]
I've checked the reason why Rome crashes here.
We have the `noAssignInExpressions` rule.
https://github.com/rome/tools/blob/22ce6c620e8912752f87cad05947f5cbc684050f/crates/rome_js_analyze/src/analyzers/suspicious/no_assign_in_expressions.rs#L107-L125

We replace the token here with `===`. However, the node remains a JsAssignmentExpression, and the subsequent `noSelfAssign` rule tries to retrieve the operator from the `JsAssignmentExpression`. This causes an issue because the node now contains `===`.

I guess that the fix is convert `JsAssignmentExpression` into `JsBinaryExpression` in `noAssignInExpressions` rule.

**Comment by Conaclos** [comment]
You should certainly rewrite your code in the following way:

```js
  res.onAborted(() =>  { aborted = true; return aborted });
```

**Comment by mrazorvin** [comment]
> You should certainly rewrite your code in the following way:
> 
> ```js
>   res.onAborted(() => { aborted = true; return aborted });
> ```
> 
> Or if you don't need to return:
> 
> ```js
>   res.onAborted(() => { aborted = true });
> ```

This is the point of my issue, I expected this suggestion from linter in console not as comment on github

**Comment by Conaclos** [comment]
The linter is not able to know if you mean `() => { aborted = true }` or `() => aborted === true`. However, the parenthesis around the assignment could suggest that the former suggestion is better than the latter. We decided to opt for the latter one when the rule was designed.

**Comment by mrazorvin** [comment]
We have big codebase where code written by different people. 
I'm okay with any linter notification that I can propagate to other team members, but when liter panic, I get no info what is wrong.

It's obviously that code should be rewritten, but I expected this information from linter

**Comment by Conaclos** [comment]
@mrazorvin We totally agree. This should be fixed by #4464 in the following release :)

**Comment by Conaclos** [comment]
For now you can disables the rule that creates the issue:

```json
{
  "linter": {
    "enabled": true,
    "rules": {
      "suspicious": {
        "noAssignInExpressions": "off"
      }
    }
  }
}

**Comment by mrazorvin** [comment]
> @mrazorvin We totally agree. This should be fixed by #4473 in the following release :)

thanks, It just not always easy to change code immediately, especially when there more of those places and you need to find them manually (there not info which file was broken)
